File: HW/HW7/MinMeetingRooms.py
def minMeetingRooms(self, intervals: List[List[int]]) -> int:
    if not intervals:
        return 0
    
    start_times = sorted([i[0] for i in intervals])
    end_times = sorted([i[1] for i in intervals])

    start_pointer = 0
    end_pointer = 0
    rooms_needed = 0

    while start_pointer < len(intervals):
        if start_times[start_pointer] >= end_times[end_pointer]:
            rooms_needed -= 1
            end_pointer += 1
        rooms_needed += 1
        start_pointer += 1

    return rooms_needed

File: HW/HW7/MinMeetingRooms.java
package HW7;

import java.util.Arrays;

public class MinMeetingRooms {
    public int minMeetingRooms(int[][] intervals) {
        if (intervals.length == 0) {
            return 0;
        }

        int[] startTimes = new int[intervals.length];
        int[] endTimes = new int[intervals.length];

        for (int i = 0; i < intervals.length; i++) {
            startTimes[i] = intervals[i][0];
            endTimes[i] = intervals[i][1];
        }

        Arrays.sort(startTimes);
        Arrays.sort(endTimes);

        int startPointer = 0;
        int endPointer = 0;
        int roomsNeeded = 0;

        while (startPointer < intervals.length) {
            if (startTimes[startPointer] >= endTimes[endPointer]) {
                roomsNeeded -= 1;
                endPointer += 1;
            }
            roomsNeeded += 1;
            startPointer += 1;
        }

        return roomsNeeded;
    }

    public static void main(String[] args) {
        MinMeetingRooms mmr = new MinMeetingRooms();

        int[][] intervals1 = {{0, 30}, {5, 10}, {15, 20}};
        System.out.println(mmr.minMeetingRooms(intervals1));

        // int[][] intervals2 = {{0, 1}, {1, 2}, {2, 3}};
        // System.out.println(mmr.minMeetingRooms(intervals2));
    }
}
